1주차 과제1 -  let을 사용하지 않고 개발하기

ChoJinmok opened the pull request:
## :: 최근 작업 주제
- [x] 기능 추가
- [x] 리뷰 반영
- [x] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표
- let을 사용하지 않고 개발하기
<br />

## :: 구현 사항 설명
- 1차 제출 - 객체로 구현
  - counter, handler를 객체로 묶어 줬습니다.
- ~~2차 제출 - reder함수, handleClick함수에 인자 절달~~
  - ~~render함수 안에 count를 const로 변수 정의~~
  - ~~handleClick함수 실행 시 인자를 전달 -> reder함수에 전달 -> count변수 변경~~
- 2차 제출 - reder함수에 인자 절달
  - render함수 안에 count를 const로 변수 정의
  - button에 click 이벤트 발생 -> reder함수에 전달 -> count변수 변경 -> 다시 렌더
- 3차 제출 - 리뷰 반영
  - 추후의 유지보수에 유리하게 handleClick함수 선언 후 사용
  - handleClick 함수 매개변수 value를 좀 더 구체적인 단어로 변경
  - render 함수 매개변수에 초기값 설정
  - 숫자 버튼들을 map해줄 때 map의 콜백함수 매개변수명 구체적 단어로 변경
  - 그 외 매개변수 명 리펙토링
- 4차 제출 - 리뷰 반영
  - 매개변수를 변수에 넣어서 사용하는 것이 아닌 바로 사용
  - 숫자 버튼 map 콜백함수의 매개변수 이름 number로 간소화
- 5차 제출 - 리뷰반영
  - no-use-before-define 제외 규칙 제거

<br />

## :: 기타 질문 및 특이 사항
- ~~객체 프로퍼티를 갱신하는 것도 재할당에 들어간다면 죄송합니다..~~
- 1차 제출시에 저는 let을 제한하신 이유가 변수가 오염될 수 있는 위험이 있어서라고 생각했습니다! 그래서 객체로 묶어줌으로서 오염 위험도를 낮춰보려고 했었습니다!
- 외부 변수 없이 함수 안에서 값을 변경할 수 있는 방법은 매개변수를 만들어서 인자로 전달받아 값을 변경할 수 있지 않을까 해서 위의 구현 사항과 같이 구현해봤습니다.
- 작업별로 commit 하는 습관이 안돼있다보니 또 합쳐서 commit 해버렸습니다.. 리뷰까지 해주셨는데 죄송합니다!! 리뷰해주신대로 작업별로 commit 하는 습관 들여보겠습니다!

ChoJinmok committed:
- 1주차 과제 1 let을 사용하지 않고 개발하기 첫번쨰 제출

재할당을 하지 않고 문제를 해결해야된다고 하셔서 하나의 객체로 묶어서 해결해봤습니다.
객체 프로퍼티를 갱신하는 것도 재할당에 들어간다면 죄송합니다..

ChoJinmok committed:
- ESLint rule 추가
index.js 상단에 function no-use-before-define 추가

daadaadaah reviewed (changes requested):
1. PR에 간단하게 작업 사항 기재해주신 점 너무 좋습니다!👍 
2. 객체는 `const` 일 때도 속성이 변경이 가능하다는 점을 이용해서 문제를 해결하려고 하셨군요!
좋은 시도였지만, `let` 을 없애려는 의도에 대해 한번 고민해보시면 좋을 것 같아요!
`let`을 없애고 외부 변수 없이 값을 직접 변경할 수 있는 방법은 없을까요? 🤔

ChoJinmok committed:
render에 매개변수 설정해서 인자 전달하여 count 변경

ChoJinmok reviewed (commented):
1. 1차 제출시에 저는 let을 제한하신 이유가 변수가 오염될 수 있는 위험이 있어서라고 생각했습니다! 그래서 객체로 묶어줌으로서 오염 위험도를 낮춰보려고 했었습니다!
2. 외부 변수 없이 함수 안에서 값을 변경할 수 있는 방법은 매개변수를 만들어서 인자로 전달받아 값을 변경할 수 있지 않을까 해서 위의 구현 사항과 같이 구현해봤습니다.

ChoJinmok committed:
코드 리펙토링 도중 handleClick함수가 필요하지 않은 걸 발견해서 지워줬습니다.

daadaadaah commented on src/index.js:
```diff
@@ -20,35 +20,26 @@ function createElement(tagName, props, ...children) {
-};
-
-function render() {
+function handleClick(value) {
```
`value` 라는 추상적인 표현보다 좀더 의도가 드러나는 단어는 없을까요? 🤔

ChoJinmok replied:
단순하게 강의에 나온 코드를 따라 쳤는데 반성하게 되네요 😭
매개변수를 number로 바꿔봐야겠네요!

daadaadaah commented on src/index.js:
```diff
@@ -20,35 +20,26 @@ function createElement(tagName, props, ...children) {
+}
+
+function render(value) {
+  const count = value || 0;
```
이 코드 대신에 다른 방법으로 value가 없을 때, count에 0을 넣어주는 방법은 없을까요? 🤔

ChoJinmok replied:
ES6 문법을 알고있었는데... 앞으로는 매개변수에 기본값 설정을 하는 습관을 들여야겠네요!!
리뷰 감사합니다!

daadaadaah reviewed (changes requested):
과제 다 수행하시고 코드 리팩토링으로 한번 체크 해보신 점 너무 좋네용!👍
다만, 지금은 render 함수밖에 없지만, handleClick 함수로 빼서 사용하는게 나중에 유지보수 측면에서 더 좋다고 생각합니다!ㅎㅎ

ChoJinmok committed:
리뷰 반영
1. 추후의 유지보수에 유리하게 handleClick함수 선언 후 사용
2. handleClick 함수 매개변수 value를 좀 더 구체적인 단어로 변경
3. render 함수 매개변수에 초기값 설정
4. 숫자 버튼들을 map해줄 때 map의 콜백함수 매개변수명 구체적 단어로 변경

ChoJinmok reviewed (commented):
리뷰 반영 완료했습니다!~🙂

ChoJinmok committed:
매개변수명 리펙토링 해주었습니다.

daadaadaah commented on src/index.js:
```diff
@@ -1,2 +1,54 @@
+  render(number);
+}
+
+function render(number = 0) {
+  const count = number;
```
27~28줄을 아래처럼 작성해주면 간단해질 것 같은데, 따로 해주신 특별한 이유가 있으신가요?ㅎㅎ 
```js
function render(count = 0) {

```

ChoJinmok replied:
매개변수 같은 경우에는 함수안에서 변경이 가능하기 때문에 변수가 오염이 될 수도 있다고 생각했습니다!  🥲

daadaadaah commented on src/index.js:
```diff
@@ -1,2 +1,54 @@
+        </button>
+      </p>
+      <p>
+        {[1, 2, 3].map((btnNumber) => (
```
27~28줄을 number를 없애고 counter를 사용하게 되면, `btnNumber` 대신 `number`를 사용할 수 있을 것 같은데, 어떻게 생각하시나요? ㅎㅎ

ChoJinmok replied:
안그래도 이 부분 수정할 때 찝찝했는데...
counter로 사용하면 훨씬 코드가 간결해지는 것 같습니다!!!👍

daadaadaah replied:
스스로 이상한 점을 느낀 거 좋은 시그널이라고 생각합니다👍

daadaadaah reviewed (changes requested):
`4. 숫자 버튼들을 map해줄 때 map의 콜백함수 매개변수명 구체적 단어로 변경` 이 작업 스스로 수정할 곳을 찾아 수정하신 점 너무 잘하셨어요!👏 항상 코드 제출 전에 한번도 검토해보는 습관 너무 좋죠!👍

1. commit 안에 어떤 작업을 했는지 기재해주셔서 좋았습니다. 좀더 작업 내역을 잘 관리하고 싶으시면, 1개의 commit에 모든 작업이 들어가는게 아니라, 작업별로 commit 남겨보세요~ㅎㅎ

ChoJinmok committed:
3차 리뷰 반영
1. 매개변수를 변수에 넣어서 사용하는 것이 아닌 바로 사용
2. 숫자 버튼 map 콜백함수의 매개변수 이름 number로 간소화

ChoJinmok reviewed (commented):
작업별로 commit 하는 습관이 안돼있다보니 또 합쳐서 commit 해버렸습니다.. 리뷰까지 해주셨는데 죄송합니다!! 리뷰해주신대로 작업별로 commit 하는 습관 들여보겠습니다!

daadaadaah commented on src/index.js:
```diff
@@ -1,2 +1,52 @@
-/* eslint-disable react/react-in-jsx-scope, react/jsx-filename-extension */
+/* eslint-disable react/react-in-jsx-scope, react/jsx-filename-extension  */
+/* eslint no-use-before-define: ["error", { "functions": false }] */
```
왜 `no-use-before-define` 룰이 사용되는지에 대해 한 번더 생각해보시면 좋을 것 같아요!
참고 : https://eslint.org/docs/rules/no-use-before-define
 lint 문제를 근본적으로 해결하려면 어떻게 해야 할까요? 🤔

ChoJinmok replied:
함수 안으로 중첩하면 해결되는 것 같습니다..!😭

ChoJinmok committed:
no-use-before-define 함수 문제 render함수에 중첩해서  해결
